randovania/server/bot.py

```diff
@@ -53,16 +56,652 @@ async def reply_for_preset(message: discord.Message, versioned_preset: Versioned
+    username = ""
+    state = AsyncRacerState.NOT_STARTED
+    vod_url = ""
+    play_timestamp = 0
+    time_timestamp = 0
+    vod_timestamp = 0
+    time = 0 # race time in seconds
+
+    def __init__(self, username):
+        self.username = username
+        self.state = AsyncRacerState.NOT_STARTED
+        self.play_timestamp = 0
+        self.time_timestamp = 0
+        self.vod_timestamp = 0
+        self.vod_url = ""
+        self.time = 0
+
+    def reset(self):
+        self.__init__(self.username)
+
+async_races = list()
+
+def async_race_idx_from_name(name):
+    i = 0
+    for race in async_races:
+        if race.name.lower() == name.lower():
+            return i
+        i = i + 1
+    return None
+
+def async_player_idx_from_name(idx, name):
+    name = name.lower()
+    i = 0
+    for player in async_races[idx].players:
+        if player.username.lower() == name:
+            return i
+        i = i + 1
+    return None
+
+def async_race_usage(is_async_race_admin):
+    if is_async_race_admin:
+        return ASYNC_RACE_USAGE_ADMIN
+    return ASYNC_RACE_USAGE
+
+def async_race_get_active_player(player_name):
+    race_idx = 0
+    for race in async_races:
+        player_idx = 0
+        for player in race.players:
+            if player.username == player_name and (player.state == AsyncRacerState.RACING or player.state == AsyncRacerState.SUBMITTED_TIME):
+                return (race_idx, player_idx)
+            player_idx = player_idx + 1
+        race_idx = race_idx + 1
+    return (None, None)
+
+def async_race_string_to_time(string):
+    try:
+        string = string.split(":")
+        return int(string[0])*60*60 + int(string[1])*60 + int(string[2])
+    except Exception:
+        return None
+
+def format_seconds_to_hhmmss(seconds):
+    hours = seconds // (60*60)
+    seconds %= (60*60)
+    minutes = seconds // 60
+    seconds %= 60
+    return "%02i:%02i:%02i" % (hours, minutes, seconds)
+
+def results_sort(x: AsyncRacePlayer):
+    return x.time
+
+async_racing_results_channel = None
+async def async_race_announce_results(race):
+    global async_racing_results_channel
+
+    if len(race.players) == 0:
+        return
+
+    # sort by time
+    players = sorted(race.players, key=results_sort, reverse=True)
+    
+    title = "%s Results" % race.name
+    embed = discord.Embed(title=title, description="Thanks for playing!", color=0x00ff00)
+    for player in players:
+        time = "DNF"
+        if player.time != DNF and player.time != 0:
+            time = format_seconds_to_hhmmss(player.time)
+
+        name = player.username.split("#")[0]
+        if player.vod_url != "":
+            vod = "[%s](%s)" % (time, player.vod_url)
+        elif time != "DNF":
+            vod = time
+        else:
+            vod = "DNF"
+
+        embed.add_field(name=name, value=vod, inline=False)
+    await async_racing_results_channel.send(embed=embed)
+
+    # notify admin of ill-play
+    ill_play = ""
```

Comment:
best practice would be to make `ill_play` a list, then you append the cases. The check at the end could be just `ill_play` (is truthy if list is not empty) and you'd do a `\n`.join(ill_play)`.